cmd/antrea-agent/agent.go

```diff
@@ -179,16 +179,26 @@ func run(o *Options) error {
 	}
 
 	var proxier k8sproxy.Provider
+
+	// list all supported EndpointSlice API version of kubernetes.
+	supportedEndpointSliceVersion := make(map[string]bool)
+	_, supportedResources, _ := k8sClient.Discovery().ServerGroupsAndResources()
+	for _, supportedResource := range supportedResources {
+		if supportedResource.Kind == "EndpointSlice" {
+			supportedEndpointSliceVersion[supportedResource.APIVersion] = true
+		}
+	}
```

Comment:
In order to avoid the situation that AntreaProxy enables EndpointSlice but Kubernetes doesn't support or disable EndpointSlice API, here we get all supported EndpointSlice version from k8sClient and store them to a map. When creating AntreaProxy, whether EndpointSlice is enabled is decided by feature gate and the supported version(only v1beta1 for now) of EndpointSlice API. @antoninbas

Reply:
I feel silently falling back to Endpoints when user turns on EndpointSlice is not necessary and might be confusing as user may expect EndpointSlice to be used.

I see @weiqiangt verified there were error logs showing that `the server could not find the requested resource (get endpointslices.discovery.k8s.io)` and I assume it will be output continuously until the API is enabled, which looks clear for troubleshooting when it's misconfigured. I also checked kube-proxy, it has same logs and behavior.

I think @antoninbas's question is to confirm the behavior in this case, guess a crash or error logs are generally acceptable as long as it's not no-op.

Reply:
I would agree we should print an error log when EndpointSlice is not enabled for the cluster or API version is not supported.

Also, I feel better to log a warning, if EndpointSlide is enabled, but AntreaProxy is not.

Reply:
> I think @antoninbas's question is to confirm the behavior in this case, guess a crash or error logs are generally acceptable as long as it's not no-op.

Yes, this is what I wanted to ensure and looks like it is the case. Maybe we can just add one sentence about this behavior in the feature-gates.md documentation: "If EndpointSlice is enabled in Antrea but the Kubernetes EndpointSlice API is disabled, error messages will be logged by Antrea and AntreaProxy will not implement ClusterIP functionality as expected".

> Also, I feel better to log a warning, if EndpointSlide is enabled, but AntreaProxy is not.
That sounds good to me, and is easy to implement

Thanks for the investigation by the way.

Reply:
Thanks for discussing this. I will remove the code above and add one sentence describing the behavior to feature-gates.md documentation.

Reply:
According to the above comments, I think the conclusion:
1. The logs in the agent can indicate that EndpointSlice API is not available for users.
2. Add description in feature-gates.md documentation about EndpointSlice: "When EndpointSlice is enabled in Antrea but EndpointSlice API is disabled in Kubernetes, there will be error messages in Antrea agent and AntreaProxy will not implement ClusterIP functionality as expected.